aiohttp/client_reqrep.py

```diff
@@ -940,19 +940,14 @@ def _response_eof(self) -> None:
         if self._closed:
             return
 
-        if self._connection is not None:
-            # websocket, protocol could be None because
-            # connection could be detached
-            if (
-                self._connection.protocol is not None
-                and self._connection.protocol.upgraded
-            ):
-                return
-
-            self._release_connection()
+        # protocol could be None because connection could be detached
```

Comment:
Can the `getattr` be avoided?

```diff
diff --git a/aiohttp/client_reqrep.py b/aiohttp/client_reqrep.py
index c0319074..2185aa85 100644
--- a/aiohttp/client_reqrep.py
+++ b/aiohttp/client_reqrep.py
@@ -941,7 +941,7 @@ class ClientResponse(HeadersMixin):
             return
 
         # protocol could be None because connection could be detached
-        protocol = getattr(self._connection, "protocol", None)
+        protocol = self._connection and self._connection.protocol
         if protocol is not None and protocol.upgraded:
             return
 
@@ -1043,7 +1043,7 @@ class ClientResponse(HeadersMixin):
         elif self._released:  # Response explicitly released
             raise ClientConnectionError("Connection closed")
 
-        protocol = getattr(self._connection, "protocol", None)
+        protocol = self._connection and self._connection.protocol
         if protocol is None or not protocol.upgraded:
             await self._wait_released()  # Underlying connection released
         return self._body
```

Reply:
Yeah, I was just thinking about that and trying to decide which is preferred.

Reply:
Looks vscode can work it out, but it looks like mypy didn't like that.

Reply:
I guess you could write it as

```python
    @property
    def _upgraded(self) -> bool:
        """Return True if connection was upgraded."""
        return (
            (connection := self._connection)
            and (protocol := connection.protocol)
            and protocol.upgraded
        )
```

Reply:
It's a bit awkward, I think. We shouldn't write lower quality code because of a bug in mypy.

Reply:
```diff
diff --git a/aiohttp/client_reqrep.py b/aiohttp/client_reqrep.py
index 2185aa85..f3e5aeb2 100644
--- a/aiohttp/client_reqrep.py
+++ b/aiohttp/client_reqrep.py
@@ -758,6 +758,15 @@ class ClientResponse(HeadersMixin):
     def __reset_writer(self, _: object = None) -> None:
         self.__writer = None
 
+    @property
+    def _upgraded(self) -> Optional[bool]:
+        """Return True if connection was upgraded."""
+        return (
+            (connection := self._connection)
+            and (protocol := connection.protocol) is not None
+            and protocol.upgraded
+        )
+
     @property
     def _writer(self) -> Optional["asyncio.Task[None]"]:
         return self.__writer
@@ -937,14 +946,8 @@ class ClientResponse(HeadersMixin):
         return self
 
     def _response_eof(self) -> None:
-        if self._closed:
+        if self._closed or self._upgraded:
             return
-
-        # protocol could be None because connection could be detached
-        protocol = self._connection and self._connection.protocol
-        if protocol is not None and protocol.upgraded:
-            return
-
         self._closed = True
         self._cleanup_writer()
         self._release_connection()
@@ -1043,8 +1046,7 @@ class ClientResponse(HeadersMixin):
         elif self._released:  # Response explicitly released
             raise ClientConnectionError("Connection closed")
 
-        protocol = self._connection and self._connection.protocol
-        if protocol is None or not protocol.upgraded:
+        if not self._upgraded:
             await self._wait_released()  # Underlying connection released
         return self._body
 
```


Another option, but I think I like the ignore better

Reply:
Hmm, the mypy developers have pointed out that if `Connection` is subclassed, then `__bool__()` could return False, in which case the mypy error is correct.

If it were made Final, then it'd be fine. But, I'm starting to think that the getattr() is the safer option.

Reply:
Somebody will subclass it so the getattr seems safer even if it's less desirable

Reply:
Also everything ran fine overnight. Logs look good. Nothing amiss

Reply:
Or, they've suggested `(conn or None) and conn.protocol`.

Reply:
The other suggestion is that we could add:
`def __bool__(self) -> Literal[True]: return True`

Which forces all subclasses to not be falsy.

Reply:
I'm going to try the latter option, as it'd make code simpler throughout the project (and we probably already made this assumption elsewhere...).